### PR TITLE
Add a --dry-run build check of cabal.project.release

### LINKED ISSUE #9601: Add a CI check of cabal.project.release?
I suspect that `cabal.project.release` is seldom touched. We might ensure that it is kept up-to-date by adding its build to CI with:

```
$ cabal build all \
  --enable-tests --enable-benchmarks --dry-run \
  --project-file=cabal.project.release \
  --index-state="hackage.haskell.org HEAD"
```

https://github.com/haskell/cabal/blob/49e3d183c0a9ac16da8d4afacc6b8a274ee67608/cabal.project.release#L1-L8

On the master branch at 49e3d183c0a9ac16da8d4afacc6b8a274ee67608:

```
$ cabal build all \
  --enable-tests --enable-benchmarks --dry-run \
  --project-file=cabal.project.release \
  --index-state="hackage.haskell.org HEAD"
Resolving dependencies...
Error: [Cabal-7107]
Could not resolve dependencies:
[__0] trying: cabal-install-3.11.0.0 (user goal)
[__1] trying: cabal-install:*test
[__2] unknown package: Cabal-described (dependency of cabal-install *test)
[__2] fail (backjumping, conflict set: Cabal-described, cabal-install, cabal-install:test)
After searching the rest of the dependency tree exhaustively,
these were the goals I've had most trouble fulfilling: cabal-install, cabal-install:test, Cabal-described
```

_Originally posted by @philderbeast in https://github.com/haskell/cabal/issues/9565#issuecomment-1881711036_

### PR DESCRIPTION
**Pull request by philderbeast**
Fixes #9601. Add a `build --dry-run` quick job for lightweight checking `cabal.project.release`. Increases duplication in `workflows/quick-jobs.yml`.

**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

File: .github/workflows/quick-jobs.yml
Comment (andreasabel) [inline .github/workflows/quick-jobs.yml:147]: Is this workaround for ghcup still needed?  The original issue is closed:
- https://github.com/actions/runner-images/issues/7061
- cf. https://github.com/herp-inc/hs-opentelemetry/commit/ef402ff9b641e74fcf86adaf38bd5ce8e224fc68

File: .github/workflows/quick-jobs.yml
Comment (andreasabel) [inline .github/workflows/quick-jobs.yml:136]: Is this always the correct path? (XDG...)

**Review by andreasabel — COMMENTED** [review]
Thanks for the workflow!

**Comment by ulysses4ever** [comment]
@philderbeast are you planning to work on this? No rush, of course.

**Review by ulysses4ever — CHANGES_REQUESTED** [review]
Since https://github.com/actions/runner-images/issues/7061 is closed, the workaround shouldn't be duplicated yet again.

**Comment by philderbeast** [comment]
Yes I am, @ulysses4ever, but have been putting all my focus on #9578, trying to get that one over the line ;-)

File: .github/workflows/quick-jobs.yml
Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:147]: Is there an example in our workflows that doesn't use this workaround (all the ones in `quick-jobs.yml` seem to)? Should we raise a separate pull request to strip these workarounds out altogether?

Comment (ulysses4ever) [reply .github/workflows/quick-jobs.yml:147]: I'd prefer this be dinner in the current PR, personally. Too granular PRs are straining our reviewing resource. And I don't feel excited about approving blind copy-paste that is even not needed anymore most likely.

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:147]: Those `sudo ...` commands are annoyingly repetitive but this is used with every other job in `quick-jobs.yml`. I saw the pattern and tried to fit in with it.

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml]: I've remove the extra duplicate of the workaround that would have been added by this new job but I've not done the others in this workflow, nor those in other workflows.

Comment (ulysses4ever) [reply .github/workflows/quick-jobs.yml]: Sounds good, thank you!

File: .github/workflows/quick-jobs.yml
Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:136]: Not sure but maybe for this job it doesn't matter. Being a `cabal build --dry-run` perhaps I don't need the cache?

Comment (ulysses4ever) [reply .github/workflows/quick-jobs.yml:136]: That's s good point.

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml]: I've removed the cache action from this job.

**Comment by philderbeast** [comment]
> Since [actions/runner-images#7061](https://github.com/actions/runner-images/issues/7061) is closed, the workaround shouldn't be duplicated yet again.

@ulysses4ever I removed the workaround. Could you please take another look at this?

File: cabal.project.release
Comment (ulysses4ever) [inline cabal.project.release:5]: I think the dev packages were omitted on purpose, and what this job should be doing is taking only the main 4 packages and dry-running a build with tests and benchmarks disabled. That's, more or less, what is happening in https://github.com/haskell/cabal/blob/master/.gitlab/ci.sh and that's what needs testing IMO.

**Review by ulysses4ever — COMMENTED** [review]
I'm sorry for the slow turnaround: I have been overwhelmed recently.

File: cabal.project.release
Comment (philderbeast) [reply cabal.project.release]: Oops sorry about that. I see my mistake, having `--enable-tests --enable-benchmarks` on the command line that overrides the project setting. I like the fish shell but will have to be more careful with [suggestions](https://fishshell.com/docs/current/interactive.html#autosuggestions) and reusing the command history.

Comment (philderbeast) [reply cabal.project.release]: And worse still, I'd copied that command line with `--enable-tests --enable-benchmarks` into the workflow!

Comment (ulysses4ever) [reply cabal.project.release]: Happens to me all the time! (I use fish too.)

File: .github/workflows/quick-jobs.yml
Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:144]: That one I put there intentionally to override the `--index-state` from the project.

https://github.com/haskell/cabal/blob/833a17deb0a5c5f99849a5b6bf06b5346105b918/cabal.project.release#L8

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:144]: For CI we want to test with `HEAD` and bump the timestamp in the project when cutting a release, wouldn't we?

Comment (ulysses4ever) [reply .github/workflows/quick-jobs.yml:144]: I see... This is maybe a matter of preference but I think if you want to test cabal.project.release, it seems strange to override anything in it. But I can see someone arguing that testing the release config against the current Hackage state may be worthy.

At the very least, please, add a comment there why this index-state flag is needed.

**Review by ulysses4ever — APPROVED** [review]

File: .github/workflows/quick-jobs.yml
Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:144]: @Mikolaj in [making a release](https://github.com/haskell/cabal/wiki/Making-a-release) should we be making mention of `cabal.project.release` and when to set its `index-state`? Do you think testing with `HEAD` but fixing `index-state` at release is the right way to go?

Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:144]: @philderbeast: well spotted, we need to do something. Actually, @chreekat, @andreabedini, could you remind me why we are now setting the Hackage index in the project file (I reviewed the PR that does, but I forgot)? That sounds like one more extra step in the release checklist, which is murderously long already.

Comment (chreekat) [reply .github/workflows/quick-jobs.yml:144]: Hey, sorry for the slow response. I've got family visiting this week. :) Let me see...

I guess there's a few things to respond to. I will put them all in this message, even though not all of them relate to the exact question @Mikolaj pinged me on.

To begin, in response to the issue itself (#9601), cabal.project.release is already tested frequently in the actual release infrastructure. By "frequently" I mean [a couple of times per day](https://gitlab.haskell.org/haskell/cabal/-/pipelines?page=1&scope=all&ref=cabal-head). I was even asked to add a badge to the README with its status, which happened in https://github.com/haskell/cabal/pull/9781. So I think cabal.project.release gets plenty of exercise. But there's still work that needs doing, as I'll explain soon.

To answer the direct question: Why does cabal.project.release set an index state? It's to reduce the pain of making releases. If the index state is constantly changing, new incompatibilities can be introduced at any moment, including the day of a release. You don't want people tinkering with the launchpad on launch day.

This is even *more* true for regular CI. The index state should be fixed so you can just test the changes to Cabal itself. Otherwise you're testing the entire wide world of Haskell in every CI run. CI would be unreliable --- a deadly sin.

So is everything just fine as it is? No. Cabal *should* be updating the index state for cabal.project.release on some kind of schedule. Ideally it should do this in isolation unrelated to any actual change in Cabal. But how often? Maybe once per release branch? Or every other..? It depends on Cabal's appetite for churn.

Another wrinkle: Cabal still has its CI split between GitHub and GitLab. Releases happen on GitLab. It may be ineffective to test a newer index state on GitHub as long as this is the case.

I have some suggestions.

First, Cabal needs to implement tests in release CI. This would be a good way to ensure that bumping the index state not only *builds*, but also *runs correctly*. I started doing that in https://gitlab.haskell.org/haskell/cabal/-/commits/b%2Ftest-bindist-tests, but it's not done.

Second, as a *post-release* step, the release manager could bump the index state in cabal.project.release. (Ideally this could be automated.) They could then wash their hands of the matter. If the bump actually broke anything, the early-warning CI would start failing, and I would get notified. (The badge would also show a problem.) I could then fix it and/or open a ticket. It wouldn't be the release manager's responsibility, and we'd have plenty of time to address the failure. The only question is how frequently to do this post-release step: every release of every minor version? Every major version? I would be fine with any schedule, but as I said, it depends on Cabal's appetite for churn.

P.S. I have focused on changes to cabal.project.release here, but presumably the rest of CI also depends on some index state...! It also needs to be updated on some schedule!

Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:144]: @chreekat thank you the exhaustive response. IMHO:

> So I think `cabal.project.release` gets plenty of exercise.

That's very true and this is relevant exercise, because it's done in the release CI. However, I think it's also worthwhile to test whether the environment GHA sets up together with `cabal.project.release` exposes any cabal problems. Also, GHA runs tests, while gitlab doesn't.

> It's to reduce the pain of making releases.

All right. That makes a lot of sense. FYI: @ffaf1 and @Kleidukos.

> This is even more true for regular CI. The index state should be fixed so you can just test the changes to Cabal itself. Otherwise you're testing the entire wide world of Haskell in every CI run. CI would be unreliable --- a deadly sin.

Hah, in that case, our GHA CI is terrible in that respect and has always been. Let's think how to improve the situation.

Note that we do want to test the entire world, because it's cabal that gets reports about the entire world failures first, a lot of the time, when users try to build things and fail. However, controlling *when* we test the entire world would be very useful. It seems introducing `cabal.project.release` to GHA lets us run at least one isolated test on GHA, which is progress and permits comparison with the other GHA jobs (and user reports).

> So is everything just fine as it is? No. Cabal should be updating the index state for cabal.project.release on some kind of schedule. Ideally it should do this in isolation unrelated to any actual change in Cabal. But how often? Maybe once per release branch? Or every other..? It depends on Cabal's appetite for churn.

Aha! That makes sense. Once per release branch, at the start of the release process, sounds reasonable, because we want to test against the libraries that has just been updated for the new GHC version that sync with (if we do, but usually we do). Your thoughts? @ffaf1, @Kleidukos? That needs to go into the Release Checklist wiki.

> It may be ineffective to test a newer index state on GitHub as long as this is the case.

Agreed but, again, we have huge gaps in testing in this respect, so this is already helping.

> First, Cabal needs to implement tests in release CI.

Spot on.

> Second, as a post-release step, the release manager could bump the index state

That has the big advantage of not blocking the release. Or is this a disadvantage? Let's discuss.

> presumably the rest of CI also depends on some index state

I'm afraid we don't have that yet. Testing with raw Hackage makes sense, given how Haskell ecosystem breakages are often reported for cabal, but we don't quite have the resources to diagnose and work around such breakages (but if we don't, who else is going to do this?). Having both fixed-index tests and newest-index tests is useful, but there's the information overload problem and CI times, so maybe limiting ourselves to fixed-index is best. Let me ask on devs chat tomorrow, because this is a hard question.

@chreekat: I appreciate you devops perspective on this very much. Thanks a lot.

@philderbeast: now that we have the clarity that the index state is actually useful, I'd propose to go ahead and merge. We'd adjust in the future depending CI and release process changes. If somebody objects, there are the 2 days of delay to signal the objections.

**Review by Mikolaj — APPROVED** [review]
LGTM

File: .github/workflows/quick-jobs.yml
Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:144]: > However, controlling when we test the entire world would be very useful. It seems introducing cabal.project.release to GHA lets us run at least one isolated test on GHA, which is progress and permits comparison with the other GHA jobs (and user reports).

I was wrong on that point, because the current PR tests with the index overriden to HEAD. GIven that it's a quick dry run, maybe it would be useful to also test with the hardwired index?

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:144]: Are you wanting both, so two runs, one with `--index-state` from `cabal.project.release` and another overriding that using `HEAD`?

Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:144]: Yes, that would be my idea. The run with the built-in `--index-state` would sometimes reveal if a breakage we are seeing comes from Hackage or from GHA changes or the very PR being tested.

**Comment by philderbeast** [comment]
Got a segfault running workflow `Validate macos-latest ghc-8.8.4`. I'll rerun it.

```
Configuring test suite 'rpmvercmp' for Cabal-tests-3..
Warning: Packages using 'cabal-version: >= 1.10' and before 'cabal-version:
3.4' must specify the 'default-language' field for each component (e.g.
Haskell98 or Haskell2010). If a component uses different languages in
different modules then list the other ones in the 'other-languages' field.
Preprocessing test suite 'rpmvercmp' for Cabal-tests-3..
Building test suite 'rpmvercmp' for Cabal-tests-3..
[1 of 1] Compiling Main             ( tests/RPMVerCmp.hs, /Users/runner/work/cabal/cabal/dist-newstyle-validate-ghc-8.8.4/build/x86_64-osx/ghc-8.8.4/Cabal-tests-3/t/rpmvercmp/build/rpmvercmp/rpmvercmp-tmp/Main.o )
Linking /Users/runner/work/cabal/cabal/dist-newstyle-validate-ghc-8.8.4/build/x86_64-osx/ghc-8.8.4/Cabal-tests-3/t/rpmvercmp/build/rpmvercmp/rpmvercmp ...
Error: cabal: Failed to build cabal-install-3.11.0.0 (which is required by
test:unit-tests from cabal-install-3.11.0.0, test:mem-use-tests from
cabal-install-3.11.0.0 and others). The build process segfaulted (i.e.
SIGSEGV).
```

**Comment by philderbeast** [comment]
Label merge+no rebase is necessary when the pull request is from an organisation.

File: .github/workflows/quick-jobs.yml
Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:146]: After a thought, I agree, bumping the index after a release is best, since it makes the already exhausting release process lighter. I've added the point to our release checklist. The counter-argument is that our releases often sync with a new GHC and we want to know if the toolchain builds with all the new versions of libraries prepared for the GHC just released on Hackage. However, our CI is almost never updated to include the new GHC (pre-alpha; so we'd need head.hackage for that) in time for the release, because too many problems emerge (so syncing with a GHC is partially decoupled with the release proper, with its out ticket for tracking the tasks, e.g., see https://github.com/haskell/cabal/issues/9729). So it's impossible to early test in CI all the new versions of packages, because we can't even test the new base. So the conservative index is just fine.

Comment (philderbeast) [reply .github/workflows/quick-jobs.yml:146]: > First, Cabal needs to implement tests in release CI.

@Mikolaj and @chreekat, are we going to enable tests in `cabal.project.release`?

https://github.com/haskell/cabal/blob/46e822152c9b3f38044870115d7a198b3a509d66/cabal.project.release#L5

Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:146]: I think @chreekat wants to do that on gitlab CI eventually. Given that our GHA CI takes very long already, maybe it's too much until we are ready to review our CI and trim down some other non-essential or near-duplicate jobs.

Comment (Mikolaj) [reply .github/workflows/quick-jobs.yml:146]: > After a thought, I agree, bumping the index after a release is best

I'm not so sure any more. We've just had a serious release issue due to that policy. Let's move the discussion to https://github.com/haskell/cabal/issues/9819.

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.12

**Pull request merged**
